### DISCUSSION TITLE
Servicelb pooling and reverse proxying the k3s control plane

### DISCUSSION THREAD

**Discussion by vasdee** [k3s-io/k3s#11431] [Q&A]
I posted this question on SO a little while ago but thought I might ask here, since this seems to be a little more activity.  I'm evaluating k3s to try and achieve network segmentation, in this case following ISA95, where a group of nodes will act as a 'sub cluster' of sorts. Each 
'sub cluster' will basically run the same core components, but will only communicate to the network segment directly northwards from the current sub-cluster....... I promise there's a good reason behind all this though, but out of scope for this question. 

There's two main components I need to get going for this, one of them is to host a mirrored registry on a 'sub cluster'. The other is to reverse proxy the control plane back to either the next higher 'sub cluster' or if already at the top of the stack, the server load balancer IP address.

So far I've managed to follow the docs to create a servicelb, that only runs on a subset of nodes, that represents 1 of the network segments. This allows me to host a container registry as a service that will provide container images for the nodes 'below'.  

My next task is to try and get use the same servicelb setup, to reverse proxy the controlplane IP/CNAME. The problem is, unlike the  registry where I can create a `type: loadbalancer` for the service and essentially target the specific nodes that are running the registry pods, `type: ExternalName` doesn't seem to want to play nice with that setup. This is where I'm getting stumped, on the surface it looks like ExternalName should do the job, but my reading has also lead me to EndpointSlices, which also look like they could do the same. Am I on the right track here, or would there be another alternate way to achieve what I'm trying to achieve? 

Original question with a pretty diagram if that helps:
https://stackoverflow.com/questions/79222201/connecting-a-k3s-agent-via-another-agents-control-plane

**Comment by brandond** [comment]
I don't think I'm really getting what you're trying to do here. Are you trying to have agents connect to the apiserver THROUGH other agents somehow? To what end? Kubernetes aside, how are you handling the K3s supervisor API?

The CNI itself also assumes a fairly flat network where all nodes are reachable to each other, even if you jurry-rig a registry and apiserver somehow, I think there are a lot of other things that are still going to bypass what you've got set up here.

Kubernetes isn't really built to tier like this.

**Comment by vasdee** [comment]
>  Are you trying to have agents connect to the apiserver THROUGH other agents

Yes, that's exactly what I'm trying to achieve here, in order to isolate a subset of nodes to only operate with certain network segments and only communicate northwards to the higher network segment

>The CNI itself also assumes a fairly flat network where all nodes are reachable to each other, even if you jurry-rig a registry and apiserver somehow, I think there are a lot of other things that are still going to bypass what you've got set up here.

This unfortunately is what I was ultimately hoping to prove/disprove as a final test in my proof-of-concept setup, once I got the control plane proxied. Nodes living within a cloud based VNET, but scattered between 3 or 4 isolated subnets with very limited firewalls between.

I was hoping based on what I had read about the hybrid/multi-cloud (https://docs.k3s.io/networking/distributed-multicloud) setups that this would be possible, without a lot of extra inter-node communications

**Reply by brandond** [reply]
No, that just uses tailscale as a VPN coordinator to build the wireguard mesh between nodes. The nodes still all end up having connectivity to each other.

Why are you trying to have agents talk through other agents, instead of just setting everything up to talk through tailscale and the servers?

**Reply by vasdee** [reply]
It's for an enterprise setup within a control systems network. ISA95 helps to isolate workloads and applications, in a nice cyber-friendly way. Also, the control systems network we are (maybe _were_ is a more realistic term now) hoping to deploy K3S is completely air-gapped, so the cloud hosted 6443 servers and registry comms need to eventually be proxied via the level 3.5 (DMZ) level nodes. Lower level nodes in the level 3 and 2 layers, can only communicate to the layers above. 

I have a better diagram of what I'm trying to achieve, which might explain it better, but I don't think I'd be allowed to post it publicly. I can send if you like

**Comment by vasdee** [comment]
Keeping this up to date in case anyone else is attempting something similar. I battled long and hard with trying to get the control plane ingressed using Traefik. In the end I swapped out traefik with nginx and it now works

Issue I'm facing now is that starting k3s throws 'unknown certificate authority' errors which I'm trying to work through now. 

```
DEBU[0002] Running k3s-cert-monitor certificate expiration check
ERRO[0002] Failed to connect to proxy. Empty dialer response  error="tls: failed to verify certificate: x509: certificate signed by unknown authority"
ERRO[0002] Remotedialer proxy error; reconnecting...     error="tls: failed to verify certificate: x509: certificate signed by unknown authority" url="wss://l35-edge.k3s.local/v1-k3s/connect"
INFO[0002] Running kube-proxy --cluster-cidr=10.42.0.0/16 --conntrack-max-per-core=0 --conntrack-tcp-timeout-close-wait=0s --conntrack-tcp-timeout-established=0s --healthz-bind-address=127.0.0.1 --hostname-override=k3s-agent-ot01 --kubeconfig=/var/lib/rancher/k3s/agent/kubeproxy.kubeconfig --proxy-mode=iptables
E0114 01:04:43.826634  197320 server.go:1051] "Failed to retrieve node info" err="Get \"https://l35-edge.k3s.local/api/v1/nodes/k3s-agent-ot01\": tls: failed to verify certificate: x509: certificate signed by unknown authority"
```
